code/modules/vending/vending.dm

```diff
@@ -652,6 +655,37 @@ ADMIN_INTERACT_PROCS(/obj/machinery/vending, proc/throw_item, proc/admin_command
 					logTheThing(LOG_STATION, usr, "vended a logged product ([product.product_name]) from [src] at [log_loc(src)].")
 				if(player_list)
 					logTheThing(LOG_STATION, usr, "vended a player product ([product.product_name]) from [src] at [log_loc(src)].")
+
+		if("o2_eject")
+			var/obj/machinery/vending/air_vendor/air_machine = src
+			if(air_machine.holding)
+				usr.put_in_hand_or_eject(air_machine.holding)
+				air_machine.holding = null
+				UpdateOverlays(null, "o2_vend_tank_overlay")
+		if("o2_insert")
+			if (istype(I, /obj/item/tank))
+				var/obj/machinery/vending/air_vendor/air_machine = src
+				air_machine.insert_tank(I, usr)
+
+		if("o2_changepressure")
+			var/obj/machinery/vending/air_vendor/air_machine = src
+			if(isnum_safe(params["pressure"]))
+				air_machine.target_pressure = clamp(params["pressure"], air_machine.min_pressure, air_machine.max_pressure)
+
+		if("o2_fill")
+			var/obj/machinery/vending/air_vendor/air_machine = src
+			if (air_machine.holding)
+				var/cost = air_machine.fill_cost()
+				if(credit >= cost)
+					src.credit -= cost
+					air_machine.fill()
+					return
+				else if(src.scan)
+					var/datum/db_record/account = FindBankAccountByName(src.scan.registered)
+					if (account && account["current_money"] >= cost)
+						account["current_money"] -= cost
+						air_machine.fill()
+						return
```

Comment:
Why not just put this on the subtype instead of casting `src`?

Reply:
At first I tried doing that, then I gave up pretty fast when I noticed hiccups with the parent's interactivity check and when I noticed that "player vending machine exclusives" were also in that same `ui_act`.

Looking into it more, the problem is that the parent `ui_act` returns true when the user is disallowed from interacting with the ui, and the children `ui_act`s return true when an interface refresh is needed. So if I were to add a second `ui_act` on the subtype, I wouldn't be able to know if the TRUE I'm getting from calling the parent is because UI interaction isn't allowed or if it is allowed, and it's that the children wants the UI refreshed.

Should I do some sort of hack where it returns something like -1 when the UI isn't interactive, or make some `ui_act_o2` method on the subtype and call it from `ui_act`? These are the alternatives I'm seeing

Reply:
Ah, well the issue is that `ui_act` is meant to return true when it "handles" a request and false otherwise, but this one is set up to return false when it fails to handle a request and true otherwise. Basically you can add `return TRUE` to all the cases on the parent where it's meant to return true and remove the `. = TRUE` to then be able to do `if (..()) return` in the child `ui_act` to check if either the UI is not interactive *or* the parent has already handled the request.